### PR TITLE
feat: implement skipping backoff strategy

### PR DESCRIPTION
**Pull request by daveleek**
# Description

Implements a backoff strategy for metrics and features endpoints based on skipping scheduled polls/posts.
Specific failed request responses (401, 403, 404, 429, 500, 502, 503, 504) increases amount of requests to skip.
10 skips in a row is maximum.
401, 403, 404 cause the skip counter to increase to 10 immediately

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Unit tests to provoke the behavior has been added

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs
Comment (sighphyre) [inline tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs:121]: Wouldn't we need one more TooManyRequests for this to check that it caps at 10? I count 10 in that list and then Ok responses

File: tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs
Comment (sighphyre) [inline tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs:163]: Up to you but I'd yeet this comment. Hopefully anyone reading this knows how loops work

```suggestion
```

File: tests/Unleash.Tests/Communication/UnleashApiClient_Metrics_Backoff_Tests.cs
Comment (sighphyre) [inline tests/Unleash.Tests/Communication/UnleashApiClient_Metrics_Backoff_Tests.cs:11]: Casing style seems to have changed here

```suggestion
    public void Should_Not_Skip_First_Call()
```

File: tests/Unleash.Tests/Communication/UnleashApiClient_Metrics_Backoff_Tests.cs
Comment (sighphyre) [inline tests/Unleash.Tests/Communication/UnleashApiClient_Metrics_Backoff_Tests.cs:151]: Just looking at the test name, I don't see a check for responses == 10? Is that intentional? Am I reading the test wrong?

File: tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs
Comment (daveleek) [reply tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs:121]: Good catch, I think I had that at some point? I'll fix! thanks!

File: tests/Unleash.Tests/Communication/UnleashApiClient_Metrics_Backoff_Tests.cs
Comment (daveleek) [reply tests/Unleash.Tests/Communication/UnleashApiClient_Metrics_Backoff_Tests.cs:11]: I blame copilot :innocent_whistle:

File: src/Unleash/Communication/UnleashApiClient.cs
Comment (sighphyre) [inline src/Unleash/Communication/UnleashApiClient.cs:27]: Name is a bit confusing to me. `featureRequestsToSkip` maybe?

**Review by sighphyre — APPROVED** [review]
Don't know the feature very well but overall I think this LGTM. I would probably wait for a review from someone who's actually implemented this one tho

File: tests/Unleash.Tests/Communication/UnleashApiClient_Metrics_Backoff_Tests.cs
Comment (daveleek) [reply tests/Unleash.Tests/Communication/UnleashApiClient_Metrics_Backoff_Tests.cs:151]: Yeah can't check it directly, so there's a loop of 10 in there along with a check that call count hasn't increased as a result of that loop - indicating it went straight to skipping 10, and then the additional invocation that should be let through the skip check and increase call count by 1

File: src/Unleash/Communication/UnleashApiClient.cs
Comment (daveleek) [reply src/Unleash/Communication/UnleashApiClient.cs:27]: Good point, yes will do! thanks

File: tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs
Comment (daveleek) [reply tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs:121]: Ok I've added an 11th bad response and slightly improved comments to show what's going on

File: tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs
Comment (chriswk) [inline tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs:140]: 55?

Comment (chriswk) [reply tests/Unleash.Tests/Communication/UnleashApiClient_Features_Backoff_Tests.cs:140]: oh, right SUM(1..10) for the skips.

**Review by chriswk — APPROVED** [review]
Nice. Looks recognizable :)

**Pull request merged**
